src/Store.ts

```diff
@@ -0,0 +1,392 @@
+import * as aspect from 'dojo-core/aspect';
+import Evented from 'dojo-core/Evented';
+import { default as has, add as hasAdd } from 'dojo-core/has';
+import { EventObject, Handle, Hash } from 'dojo-core/interfaces';
+import * as lang from 'dojo-core/lang';
+import Promise from 'dojo-core/Promise';
+import Filter from './Filter';
+import * as dstore from './interfaces';
+import QueryMethod, { QueryMethodArgs } from './QueryMethod';
+
+// Detect __proto__, and avoid using it on Firefox, as they warn about
+// deoptimizations. The watch method is a clear indicator of the Firefox
+// JS engine.
+hasAdd('object-proto', Boolean((<any> {})['__proto__']) && !(<any> {}).watch);
+const hasProto = has('object-proto');
+
+export interface NewableStoreModel {
+	new (args?: {}): any;
+	createSubclass?: (arg: any[]) => any;
+	prototype: any;
+}
+
+export interface StoreArgs {
+	Model?: NewableStoreModel;
+}
+
+export default class Store<T> extends Evented implements dstore.Collection<T>, Hash<any> {
+	[ index: string ]: any;
+	/**
+	 * Creates an object, throws an error if the object already exists
+	 *
+	 * @param object The object to store.
+	 * @param directives Additional directives for creating objects.
+	 *
+	 * @return The object that was stored, with any changes that were made by
+	 * the storage system (like generated id)
+	 */
+	add: (object: any, directives?: dstore.PutDirectives) => any;
```

Comment:
This should return a promise (just like `put`).  We might also argue that `add` and `put` should return specifically `Promise<T>`.

IIRC, the return value of `add` and `put` has historically been a little more flexible and vague than I'd like, but at the very least, the docs are clear on this - it's supposed to return the item that was stored.